### PR TITLE
Add BridgeTowerForContrastiveLearning

### PR DESCRIPTION
**Pull request by abhiwand**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 1]
```diff
@@ -1698,3 +1732,136 @@ def forward(
...
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> texts = "An image of two cats chilling on a couch"
+
+        >>> processor = BridgeTowerProcessor.from_pretrained("BridgeTower/bridgetower-large-itm-mlm-itc")
+        >>> model = BridgeTowerForITC.from_pretrained("BridgeTower/bridgetower-large-itm-mlm-itc")
+        >>> outputs = model(**inputs, output_hidden_states=True)
+        ```"""
+        assert output_hidden_states, 'output_hidden_states should be set to True for BridgeTowerForITC'
```
Comment (sgugger) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:1803]: No new asserts in the code base, please use a test and raise an appropriate error.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 2]
```diff
@@ -1698,3 +1732,136 @@ def forward(
...
+
+        image_embeds = (
+            self.bridgetower.cross_modal_image_transform(image_embeds_with_ln)
+            + image_token_type_embeddings
+        )
+
+        # normalized features
+        text_embeds = normalize(self.itc_text_head(text_embeds[:,0,:]), dim=-1, p=2)
+        image_embeds = normalize(self.itc_image_head(image_embeds[:,0,:]), dim=-1, p=2)
+        cross_embeds = normalize(self.itc_cross_modal_head(pooler_output), dim=-1, p=2)
```
Comment (sgugger) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:1839]: Let's use nn.functional.normalize here, so it's clear to the user where the normalize comes from.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 3]
```diff
@@ -57,6 +58,10 @@
 ]
 
 
+def contrastive_loss(logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
+    return nn.functional.cross_entropy(logits, labels)
```
Comment (sgugger) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:62]: Since it's just one line, let's not define a function for this and just use cross entropy.

File: src/transformers/models/bridgetower/modeling_bridgetower.py [hunk 4]
```diff
@@ -23,6 +23,7 @@
 import torch.utils.checkpoint
 from torch import nn
 from torch.nn import CrossEntropyLoss
+from torch.nn.functional import normalize
```
Comment (sgugger) [inline src/transformers/models/bridgetower/modeling_bridgetower.py:26]: As said below, let's not import this and use it via `nn.functional.normalize`.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this new model. Make sure to fix all the auqlity tests by running `make style` locally and `make repo-consistency`.

The new model should be added in the common tests in aprticular.

**Comment by tileintel** [comment]
@sgugger We have addressed all of your review in the latest commit. We have also added tests for BridgeTowerForContrastiveLearning. Would you please review the latest commit? We are looking forward to having this PR merged into main.
Thanks

**Review by sgugger — APPROVED** [review]
Thanks for iterating on this!

**Comment by amyeroberts** [comment]
Thanks for adding this model @abhiwand @tileintel !  Great to see it merged into the library :) 

For the model tests, some of the configuration values in `BridgeTowerModelTester` result in large models being created and used in the test suite e.g. `vocab_size = 50265` set [here](https://github.com/huggingface/transformers/blob/bcc8d30affba29c594320fc80e4a4422fb850175/tests/models/bridgetower/test_modeling_bridgetower.py#L97), which results in periodic OOM errors in the CI runs. 

Could you add a follow up PR for `BridgeTowerModelTester` and `BridgeTowerModelTest` to have smaller default values to create lighter tests?  A good reference for this would be [CLIP](https://github.com/huggingface/transformers/blob/main/tests/models/clip/test_modeling_clip.py). Ideally we would also have a similar structure of test classes for the different modalities i.e. `BridgeTower[Text|Vision]ModelTest(er)`.
